DM-31767: Add ATCameraDewar rule

r-owen opened the pull request.

patrickingraham reviewed (commented):
I looked primarily at the ATCamera example, specifically the default values in the schema.
The primary issue was understanding what is meant by "hysteresis"

r-owen commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,420 @@
+
+
+class ATCameraDewar(watcher.BaseRule):
+    """Monitor ATCamera dewar, including CCD temperature and vacuum.
```
Well that is why I said "including", but I'll reword it.

r-owen commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,420 @@
+                max_warning_vacuum:
+                    description: Maximum dewar vacuum (Torr) above which warning.
+                    type: number
+                    default: 1.0e-6
```
Do you want both a warning level and a serious level? It's totally up to you.

Also: it can easily be changed in the watcher config file, but I'd just as soon make the defaults as reasonable as I can.

r-owen commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,420 @@
+                    description: Minimum CCD temperature (C) below which serious.
+                    type: number
+                min_critical_ccd_temp:
+                    description: Minimum CCD temperature (C) below which critical.
```
Sure. Do you also want serious or critical default levels for cold plate or cryo head?

patrickingraham replied:
Please.
Unfortunately I don't really know the values for these!
From a random sampling of data in chronograf, the cryohead can probably move by +/- 15 degrees, the cold plate by 8. These are somewhat of a wag, but we can revise them later.

r-owen replied:
(I thought I already responded to this, but I don't see it here, so I'll try again)...
We talked off-line and came up with some values which I implemented here and describe in the Jira ticket.

r-owen commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,420 @@
+                    type: number
+                    default: 180
+                    exclusiveMinimum: 0
+                temperature_hysteresis:
```
That's fair. I added the following to the description of each hysteresis:
```
The amount by which a value must improve past an
alarm threshold before the alarm severity decreases.
```
To put it another way: the thresholds set when an alarm goes off, the hysteresis determines when it goes off again. This is to avoid small fluctuations right at a level from making an alarm stutter on and off.

patrickingraham replied:
That's great. Thanks for the explanation.
Hysteresis feels like the incorrect term however. 
Maybe  "improvement_condition", "delta" or "severity_delta" depending on the length.
You're call. I'm not sure there is an easy answer here.

r-owen replied:
I am already using "hysteresis" to mean exactly this for all ESS-based rules and in the ThresholdHandler, so it'll be a big change. It's one I'm willing to consider, but I'd like to do it on a separate ticket because it involves many existing rules.

r-owen commented:
I expanded the alarm doc string and added the levels you requested.

I also made all thresholds "nullable" so you can specify "null" in a config file to disable that a threshold that has a default value. An example where that helps is if the alarm has serious threshold for a measurement, but you want a critical threshold instead: specify null for the serious threshold and the desired value for the critical threshold.

Finally I added a larger suite of bad yaml files to expand the validation test coverage.

I would like to add a second threshold for cold plate and cryo head (serious or critical) if you think that is reasonable and can suggest values.

patrickingraham commented on tests/data/config/atcamera_dewar/README.txt:
```diff
@@ -0,0 +1,6 @@
+Configuration files for the ATCameraDewar rule unit tests.
```
I like what you did here. We should make this common practice to have an "index" of the different test configs.

r-owen replied:
I do this for most rules (especially the rules with complicated configuration).

patrickingraham commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,458 @@
+            Item is being measured. One of:
+            "ccd_temp", "cold_plate_temp", "cryo_head_temp", "vacuum".
+        big_is_bad : bool
+            Are larger values bad?
```
I don't quite understand this. Can you maybe just add a sentence explaining this?

r-owen replied:
I reworded it to:
```
big_is_bad : bool
    True if measured values larger than the specified levels are bad.
    False if measured values smaller than the specified levels are bad.
```
which more closely resembles the doc string for ThresholdHandler (which I also tweaked a bit, but it was already a lot clearer than what I wrote here).

patrickingraham commented on python/lsst/ts/watcher/rules/atcamera_dewar.py:
```diff
@@ -0,0 +1,458 @@
+            oldest_vacuum_tai = curr_tai - self.config.vacuum_window
+            oldest_tai = min(oldest_temp_tai, oldest_vacuum_tai)
+
+            while self.data_queue:
```
does this loop need to run this fast? Will it chew through an unnecessary amount of CPU cycles?

r-owen replied:
I'm not sure I understand your objection. `__call__` is called whenever a new `vacuum` topic is read. This adds a new data point to the data queue, and the data is processed. Processing includes:
* Purge old data (that's what happens in this loop).
* Compute medians of what's left.
* Pass the medians to the threshold handler to see which alarm conditions to report, if any.

patrickingraham reviewed (approved):
Much better thanks.
I made a few comments, please review them before merging.
